### PR TITLE
[ADAM-843] Aggressively project out metadata fields.

### LINKED ISSUE #843: Provide selective predicates and projections for transform
For many of the `transform` stages (e.g., BQSR, indel realignment, markdups), we don't touch most record group fields. Additionally, we don't modify unmapped reads, so we can throw those away.

### PR DESCRIPTION
**Pull request by fnothaft**
Resolves #843. I have tested this on our cluster and seen a 2.5x performance improvement for INDEL realignment and MarkDups, and a 1.5x performance improvement for BQSR. Additionally, we see approximately a 3-4x reduction in shuffle size.

### CODE REVIEW

File: adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala [hunk 1]
```diff
@@ -159,8 +165,41 @@ class Transform(protected val args: TransformArgs) extends BDGSparkCommand[Trans
         sc.loadFastq(args.inputPath, Option(args.pairedFastqFile), Option(args.fastqRecordGroup), stringency)
       } else if (args.forceLoadIFastq) {
         sc.loadInterleavedFastq(args.inputPath)
-      } else if (args.forceLoadParquet) {
-        sc.loadParquetAlignments(args.inputPath)
+      } else if (args.forceLoadParquet ||
+        args.limitProjection ||
+        args.useAlignedReadPredicate) {
```
Comment (heuermh) [inline adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala]: if either of these are present along with ```-force_load_[something other than parqet]```, should the be a warning?

Comment (fnothaft) [inline adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala]: Agreed, not a bad idea. I've added logging for this at https://github.com/bigdatagenomics/adam/pull/844/files#diff-1ee985bcfcfad885cfc143944371af07R161.

Comment (ryan-williams) [inline adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala]: is it weird to have `args.limitProjection || args.useAlignedReadPredicate` cause interpretation of a file as parquet? would we rather throw an error if someone tries to load a `.bam` file with one of these flags? I'm not sure, just wondering

File: adam-core/src/main/scala/org/bdgenomics/adam/projections/AlignmentRecordField.scala [hunk 2]
```diff
@@ -28,5 +28,5 @@ import org.bdgenomics.formats.avro.AlignmentRecord
...
+  val contig, start, end, mapq, readName, sequence, mateAlignmentStart, cigar, qual, recordGroupId, recordGroupName, readPaired, properPair, readMapped, mateMapped, readNegativeStrand, mateNegativeStrand, firstOfPair, secondOfPair, primaryAlignment, failedVendorQualityChecks, duplicateRead, mismatchingPositions, attributes, recordGroupSequencingCenter, recordGroupDescription, recordGroupRunDateEpoch, recordGroupFlowOrder, recordGroupKeySequence, recordGroupLibrary, recordGroupPredictedMedianInsertSize, recordGroupPlatform, recordGroupPlatformUnit, recordGroupSample, mateContig, origQual, supplementaryAlignment, secondaryAlignment = SchemaValue
```
Comment (ryan-williams) [inline adam-core/src/main/scala/org/bdgenomics/adam/projections/AlignmentRecordField.scala]: worth wrapping this line? hard to tell what changed

File: adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala (hunk 1 above)
Comment (fnothaft) [inline adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala]: I changed this to throw an IAE.

File: adam-core/src/main/scala/org/bdgenomics/adam/projections/AlignmentRecordField.scala (hunk 2 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/projections/AlignmentRecordField.scala]: If you wrap the line, the `scripts/format-source` script helpfully unwraps the line. Sigh!

**Comment by heuermh** [comment]
Thanks!

**Pull request merged**